<a href="https://colab.research.google.com/github/MeganKom/CS167Code/blob/main/Notebook5_KomFinnished.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Notebook \#5: SVM, Perceptron, SGD, and PCA
**Due:** Thursday April 8th, by 11:59pm

Name:

**Exercises**: \\
Build a default model for each of the following and then build a model with one parameter tuned:
1. Support Vector Machines
2. Perceptron
3. Stochastic Gradient Descent
4. Run a PCA on the best model
5. Answer the conclusion questions


In [2]:
from google.colab import drive
import pandas as pd
drive.mount('/content/drive')
data= pd.read_csv('/content/drive/MyDrive/breast-cancer-wisconsin-data.csv')
data.head()

Mounted at /content/drive


id diagnosis  ...  fractal_dimension_worst  Unnamed: 32
0    842302         M  ...                  0.11890          NaN
1    842517         M  ...                  0.08902          NaN
2  84300903         M  ...                  0.08758          NaN
3  84348301         M  ...                  0.17300          NaN
4  84358402         M  ...                  0.07678          NaN

[5 rows x 33 columns]

In [16]:

from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix
from sklearn.svm import SVC
from sklearn.preprocessing import StandardScaler
from sklearn import metrics
from sklearn.decomposition import PCA


predictors = ['radius_mean', 'texture_mean', 'perimeter_mean',
       'area_mean', 'smoothness_mean', 'compactness_mean', 'concavity_mean',
       'concave points_mean', 'symmetry_mean', 'fractal_dimension_mean',
       'radius_se', 'texture_se', 'perimeter_se', 'area_se', 'smoothness_se',
       'compactness_se', 'concavity_se', 'concave points_se', 'symmetry_se',
       'fractal_dimension_se', 'radius_worst', 'texture_worst',
       'perimeter_worst', 'area_worst', 'smoothness_worst',
       'compactness_worst', 'concavity_worst', 'concave points_worst',
       'symmetry_worst', 'fractal_dimension_worst']

#predictors = data.columns.drop('diagnosis') #gets all of the columns except the diagnosis
target = "diagnosis"
train_data, test_data, train_sln, test_sln = train_test_split(data[predictors], data[target], test_size = 0.2, random_state=41)
#Normalize Data
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
scaler.fit(train_data)
train_data = scaler.transform(train_data)
test_data = scaler.transform(test_data)

## Exercises:
1. Support Vector Classifier: \\
  a. Default \\
  b. Tuned

In [8]:
#Default SVC
svc = SVC()
svc.fit(train_data,train_sln)
svc_predictions = svc.predict(test_data)
print("Default SVC: ", accuracy_score(test_sln,svc_predictions))

vals = data[target].unique()
conf_mat = metrics.confusion_matrix(test_sln, svc_predictions, labels=vals)
print(pd.DataFrame(conf_mat, index = "T" + vals, columns = "Pred" +vals))


Default SVC:  0.9824561403508771
    PredM  PredB
TM     39      1
TB      1     73


In [9]:
#Tuned SVC:
svc_tuned = SVC(C=2, max_iter=-1, random_state=1)
svc_tuned.fit(train_data,train_sln)
svc_tuned_predictions = svc_tuned.predict(test_data)
print("Tuned SVC: ", accuracy_score(test_sln,svc_tuned_predictions))

conf_mat = metrics.confusion_matrix(test_sln, svc_tuned_predictions, labels=vals)
print(pd.DataFrame(conf_mat, index = "T" + vals, columns = "Pred" +vals))

Tuned SVC:  0.9912280701754386
    PredM  PredB
TM     40      0
TB      1     73


2. Perceptron: \\
  a. Default \\
  b. Tuned

In [10]:
#Default Perceptron
from sklearn.linear_model import Perceptron

perc = Perceptron()
perc.fit(train_data,train_sln)
perc_predictions = perc.predict(test_data)
print("Default Perceptron: ", accuracy_score(test_sln,perc_predictions))

conf_mat = metrics.confusion_matrix(test_sln, perc_predictions, labels=vals)
print(pd.DataFrame(conf_mat, index = "T" + vals, columns = "Pred" +vals))

Default Perceptron:  0.956140350877193
    PredM  PredB
TM     38      2
TB      3     71


In [11]:
#Tuned Perceptron

perc_tuned = Perceptron(alpha=.1, max_iter=10)
perc_tuned.fit(train_data,train_sln)
perc_tuned_predictions = perc_tuned.predict(test_data)
print("Tuned Perceptron: ", accuracy_score(test_sln,perc_tuned_predictions))

conf_mat = metrics.confusion_matrix(test_sln, perc_tuned_predictions, labels=vals)
print(pd.DataFrame(conf_mat, index = "T" + vals, columns = "Pred" +vals))

Tuned Perceptron:  0.9736842105263158
    PredM  PredB
TM     38      2
TB      1     73


/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_stochastic_gradient.py:557: ConvergenceWarning: Maximum number of iteration reached before convergence. Consider increasing max_iter to improve the fit.
  ConvergenceWarning)


3. Stochastic Gradient Descent: \\
  a. Default \\
  b. Tuned

In [21]:
#Default SGD
from sklearn.linear_model import SGDClassifier

sgd = SGDClassifier()
sgd.fit(train_data,train_sln)
sgd_predictions = sgd.predict(test_data)
print("Default SGD Classifier: ", accuracy_score(test_sln,sgd_predictions))

conf_mat = metrics.confusion_matrix(test_sln, sgd_predictions, labels=vals)
print(pd.DataFrame(conf_mat, index = "T" + vals, columns = "Pred" +vals))

Default SGD Classifier:  0.9912280701754386
    PredM  PredB
TM     40      0
TB      1     73


In [22]:
#Tuned SGD

sgd_tuned = SGDClassifier(alpha=.000001, max_iter=1000)
sgd_tuned.fit(train_data,train_sln)
sgd_tuned_predictions = sgd_tuned.predict(test_data)
print("Tuned SGD Classifier: ", accuracy_score(test_sln,sgd_tuned_predictions))

conf_mat = metrics.confusion_matrix(test_sln, sgd_tuned_predictions, labels=vals)
print(pd.DataFrame(conf_mat, index = "T" + vals, columns = "Pred" +vals))

Tuned SGD Classifier:  0.9912280701754386
    PredM  PredB
TM     40      0
TB      1     73


3. Now take the model that performed the best above and run a **principal component analysis** on it. 

In [24]:
#PCA on best model
from sklearn.feature_selection import SelectKBest
from sklearn.decomposition import PCA

selector = SelectKBest(k=2)
selector.fit(train_data,train_sln)
print("Scores for each feature")
print(selector.scores_)

print("This is what the training data looks like after transformation.")
train_transformed = selector.transform(train_data)

test_transformed = selector.transform(test_data)
svc_tuned2 = SVC()
svc_tuned2.fit(train_transformed,train_sln)
svc_tuned2_predictions = svc_tuned2.predict(test_transformed)
print("SVC: ", accuracy_score(test_sln,svc_tuned2_predictions))

extractor = PCA(n_components=2, whiten=True)
extractor.fit(train_data)
print("variance/importance of each component")
print(extractor.explained_variance_ratio_)

Scores for each feature
[4.86076047e+02 8.05159650e+01 5.25872739e+02 4.51511787e+02
 8.77863019e+01 2.50000183e+02 4.31027328e+02 6.80719655e+02
 6.74204085e+01 3.72034958e-03 2.58731679e+02 2.84635675e-01
 2.42251517e+02 3.10650829e+02 2.10049065e+00 3.40257010e+01
 3.45760847e+01 9.09895927e+01 2.76246747e-01 2.60066487e+00
 6.54890146e+02 1.07003275e+02 6.86666761e+02 5.21809597e+02
 1.19570338e+02 2.44872777e+02 3.49224673e+02 7.65597537e+02
 1.17519232e+02 6.09183202e+01]
This is what the training data looks like after transformation.
SVC:  0.9649122807017544
variance/importance of each component
[0.45115158 0.18751237]


In [25]:
print("Before transformation")
print(train_data[0:4])

train_transformed = extractor.transform(train_data)
print("After transformation")
print(train_transformed[0:4])

Before transformation
[[-0.36498729 -0.83463794 -0.37044213 -0.42265468 -0.41437387 -0.40892601
  -0.38027127 -0.45643791 -0.37976688 -0.53426485 -1.00722501 -1.35166219
  -0.91826389 -0.7581854  -0.70035103 -0.40772137  0.05810419 -0.5528159
  -0.53017844 -0.58567284 -0.45365506 -1.06177658 -0.41985334 -0.48882443
  -0.00381561  0.0295832   0.34707213 -0.25967334  0.17192889 -0.3532633 ]
 [ 0.05208811 -0.26700868 -0.01772895 -0.04633291 -2.19428762 -0.99434456
  -0.81174734 -0.90244966 -0.63965204 -1.01238652 -0.67665163 -1.00795366
  -0.64856865 -0.50675818 -1.27321035 -0.64147195 -0.76808847 -1.08006164
  -1.19358917 -0.71798042 -0.00465333 -0.09398457 -0.03673001 -0.10175369
  -1.67045714 -0.25113922 -0.56957017 -0.60430302 -0.44112743 -0.4234167 ]
 [-0.12582517 -1.43250975 -0.15915297 -0.23777394  0.0233809  -0.32406048
  -0.63029624 -0.59123606 -1.66062944  0.09591998 -0.7583664  -1.45373321
  -0.80482984 -0.59582881 -0.94912797 -0.46406697 -0.55665105 -0.84054288
  -0.83657296 -

In [27]:
test_transformed = extractor.transform(test_data)
svc_tuned3 = SVC(C=2, max_iter=-1, random_state=1)
svc_tuned3.fit(train_transformed,train_sln)
svc_tuned3_predictions = svc_tuned3.predict(test_transformed)
print("SVC Tuned with Principle Component Analysis: ", accuracy_score(test_sln,svc_tuned2_predictions))

#conf_mat = metrics.confusion_matrix(test_sln, svc_tuned_predictions, labels=vals)
#print(pd.DataFrame(conf_mat, index = "T" + vals, columns = "Pred" +vals))

SVC Tuned with Principle Component Analysis:  0.9649122807017544


## Conclusions:
a. Which model performed the best? Why do you think this model performed the best? \\
b. What effect did the principal component analysis have on the accuracy of your best model? Why do you think this is the case?

a. Before I tuned the models, the SGD model performed the best with an accuracy of .99122 and next was SVC followed by Perception. The SGD model probably performed the best because the model linearly separated the data the best and the learning rate was very small.

.9824=SVC


.956=Perception


.99122=SGD

b. The principal component analysis helps get the best results by identifying to me which new composite features may more directly identify the signal and minimize noise in the data. This is the case becuase the highest variance is the first principal component, minimize information loss, and each new attribute is linearly uncorrelated to the others. 